include/ptp_clock.h

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright (c) 2017 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef _PTP_CLOCK_H_
+#define _PTP_CLOCK_H_
+
+#include <stdint.h>
+#include <device.h>
+#include <misc/util.h>
+#include <net/gptp.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/*
+ * Special alignment is needed for ptp_clock which is stored in
+ * a ptp_clock linker section if there are more than one ptp clock
+ * in the system. If there is only one ptp clock,
+ * then this alignment is not needed, unfortunately this cannot be
+ * known beforehand.
+ *
+ * The ptp_clock struct needs to be aligned to 32 byte boundary,
+ * otherwise the __ptp_clock_end will point to wrong location and ptp_clock
+ * initialization done in ptp_clock_init() will not find proper values
+ * for the second interface.
+ *
+ * So this alignment is a workaround and should eventually be removed.
+ */
+#define __ptp_clock_align __aligned(32)
+
+struct ptp_clock;
+
+struct ptp_clock_driver_api {
+	int (*set)(struct ptp_clock *clk, struct net_ptp_time *tm);
+	int (*get)(struct ptp_clock *clk, struct net_ptp_time *tm);
+	int (*adjust)(struct ptp_clock *clk, int increment);
+	int (*rate_adjust)(struct ptp_clock *clk, float ratio);
+};
+
+/**
+ * @brief PTP Clock structure
+ *
+ * Used to handle a ptp clock on top of a device driver instance.
+ * There can be many ptp_clock instance against the same device.
+ */
+struct ptp_clock {
+	/** The actually device driver instance the ptp_clock is related to */
+	struct device *dev;
+
+	/** API for the ptp clock. */
+	struct ptp_clock_driver_api const *api;
+} __ptp_clock_align;
+
+#define PTP_CLOCK_GET_NAME(dev_name, sfx) (__ptp_clock_##dev_name##_##sfx)
+#define PTP_CLOCK_GET(dev_name, sfx)					\
+	((struct ptp_clock *)&PTP_CLOCK_GET_NAME(dev_name, sfx))
+
+#define PTP_CLOCK_INIT(dev_name, sfx, api)				\
+	static const struct ptp_clock					\
+			(PTP_CLOCK_GET_NAME(dev_name, sfx)) __used	\
+	__attribute__((__section__(".ptp_clock.data"))) = {		\
+		.dev = &(DEVICE_NAME_GET(dev_name)),			\
+		.api = &(api),						\
+	}								\
+
+/* PTP clock initialization macros */
+
+#define PTP_CLOCK_DEVICE_INIT(dev_name, api)				\
```

Comment:
Where is this variant needed/going to be used?

Reply:
Normally this macro is used. The instance macro can be used if there are multiple instances of this driver.

Reply:
get rid of these macros.

Up to the code using ptp device to use device_get_binding().

And I don't see the use for that ptp clock structure either: device has a pointer for holding api structure.

Reply:
> Up to the code using ptp device to use device_get_binding().

I am not sure I follow you here. The macros are used in ethernet driver to setup the clock. These macros are similar as what ETH_NET_DEVICE_INIT() is doing.

Reply:
you are creating a new device driver api, and as such it needs to follow how it's done there.
You should be using struct device * as first param, no specific one. Also, if it's a device, then its API is hold in device->api, why having a specific place-holder, with its own section? (just use DEVIC_INIT()) etc...

Reply:
> why having a specific place-holder, with its own section?

The gptp code does not have direct access to PTP device, it only knows the ethernet device. That is the reason we have this PTP_CLOCK_INIT() that ties the ethernet and ptp_clock devices together. If you have some other good way to solve this, please tell me as I am a bit lost what you mean by the latest comment.